examples/src/demo_client.rs

```diff
@@ -0,0 +1,154 @@
+    NodeReadCausalRequest, PublicKey, ReadCausalRequest, RemoveCollectionsRequest, RoundsRequest,
+    ValidatorData,
+};
+
+pub mod narwhal {
+    tonic::include_proto!("narwhal");
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /*
+
+    println!("\n******************************** Configuration Service ********************************\n");
+    let mut client = ConfigurationClient::connect("http://127.0.0.1:8000").await?;
+    let stake_weight = 1;
+    let address = MultiAddr {
+        address: "/ip4/127.0.0.1".to_string(),
+    };
+
+    println!("\n---- Test NewNetworkInfo endpoint ----\n");
+    let request = tonic::Request::new(NewNetworkInfoRequest {
+        epoch_number: 0,
+        validators: vec![ValidatorData {
+            public_key: Some(PublicKey {
+                bytes: public_key.clone(),
+            }),
+            stake_weight,
+            address: Some(address),
+        }],
+    });
+
+    let response = client.new_network_info(request).await;
+
+    println!("NewNetworkInfoResponse={:?}", response);
+
+    */
+
+    println!(
+        "\n******************************** Proposer Service ********************************\n"
+    );
+    let mut client = ProposerClient::connect("http://127.0.0.1:8000").await?;
+    let public_key = base64::decode("Zy82aSpF8QghKE4wWvyIoTWyLetCuUSfk2gxHEtwdbg=").unwrap();
+    let gas_limit = 10;
+
+    println!("\n1) Retrieve the range of rounds you have a collection for\n");
+    println!("\n---- Use Rounds endpoint ----\n");
+
+    let request = tonic::Request::new(RoundsRequest {
+        public_key: Some(PublicKey {
+            bytes: public_key.clone(),
+        }),
+    });
+
+    println!("RoundsRequest={:?}\n", request);
+
+    let response = client.rounds(request).await;
+
+    println!("RoundsResponse={:?}\n", response);
+
+    let rounds_response = response.unwrap().into_inner();
+    let oldest_round = rounds_response.oldest_round;
+    let newest_round = rounds_response.newest_round;
+    let round = oldest_round + 1;
+    println!("\n2) Find collections from earliest round and continue to add collections until gas limit is hit\n");
+    println!("\n---- Use NodeReadCausal endpoint ----\n");
+
+    let mut collection_ids: Vec<CertificateDigest> = vec![];
+    while round < newest_round && collection_ids.len() < gas_limit {
+        let request = tonic::Request::new(NodeReadCausalRequest {
+            public_key: Some(PublicKey {
+                bytes: public_key.clone(),
+            }),
+            round,
+        });
+
+        println!("NodeReadCausalRequest={:?}\n", request);
+
+        let response = client.node_read_causal(request).await;
+
+        println!("NodeReadCausalResponse={:?}\n", response);
+
+        let node_read_causal_response = response.unwrap().into_inner();
+
+        if collection_ids.len() + node_read_causal_response.collection_ids.len() <= gas_limit {
```

Comment:
I know this is for demonstration, but I would like to outline that consecutive read_causal requests, for example:
* `read_causal(round=1)`
* `read_causal(round=2)` 
* `read_causal(round=3)`
....

it's possible to contain overlaps on their responses, meaning `collection_ids` that exist on multiple responses. Since we are doing a BFS and a certificate we start from might refer to a parent certificate of the previous round of ours, we always have the chance of  referring to same collection ids.

That being said, we might want to show that we are doing some deduping and we only take into account the "unique" collection ids.

CC @huitseeker

Reply:
Also, in practice, wouldn't the consensus layer want to fetch the actual transactions to figure out the gas cost? I mean, I would imagine in high level that they want to do something like (pseudo code):
````
round  = rounds().oldest_round;
newest_round = rounds().newest_round;

while round <= newest_round 
{
    let collection_ids = read_causal(round);

    // keep non overlapping collection ids
    collection_ids = dedupe_collection_ids(collection_ids, all_collection_ids);

    // fetch the actual transactions so we can
    // calculate the gas will need before proposing

    let transactions = get_collections(collection_ids);

    // calculate the gas for each transaction and return
    // the ones that should be included so we stay under
    // the limit - or whatever logic consensus wants to implement.
   
    let (collection_ids_to_include, gas) = collections_under_gas_limit(transactions, total_gas, gas_limit);
   
     total_gas += gas;
     all_collection_ids.extend(collection_ids_to_include);

     if collection_ids_to_include.len() != collection_ids.len() {
         // not all collections included, meaning we reached (or we would reach if included more)
         // the gas limit. Now abort.
         break;
     }
     
     round += 1;
}
````

Reply:
@arun-koshy @akichidis 
- it's expected for `read_causal(round=n+1)` to indeed include the `read_causal(round=n)`,
- we should indeed show deduplication,
- as far as gas cost is concerned, the fancy calculations should only exist in comments, saying "PLACE YOUR EXECUTION AND GAS ESTIMATION HERE"
- I suggest that to estimate gas we simply emulate it by saying each transaction costs one unit of gas. 
- this does force us to issue a `get_collection` and compute on its results, because the info of the $ of transactions is hidden in the batches and  not accessible from a header.